transport: validate http 200 status for responses

JNProtzman opened the pull request:
Per the [spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses) the transport should be verifying an http 200 status per request.

[Issue discussion](https://github.com/grpc/grpc-go/issues/4451)
[Initial discovery in PR](https://github.com/grpc/grpc-go/pull/3313#discussion_r627578993)

RELEASE NOTES:
* transport: fail RPCs without HTTP status 200 (OK), according to the gRPC spec

dfawley reviewed (commented):
Looks good; can you add a test either in `transport` or `test/end2end_test.go` where everything is successful except for the status?

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2045,6 +2047,25 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
+			},
+			wantStatus: status.New(
+				codes.Unavailable,
+				"Gateway Timeout: HTTP status code 504; transport: missing content-type field",
```
`transport: missing content-type field` looks incorrect, since a valid `content-type` header is present.

Also maybe the main part of the error message can be clearer?  E.g.

`unexpected HTTP status code received from server: 504 (Gateway Timeout)`

JNProtzman replied:
That does seem wrong. I've updated the messaging!

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2018,7 +2020,7 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
 			wantStatus: status.New(
 				codes.Internal,
-				": HTTP status code 0; transport: received the unexpected content-type \"application/json\"",
+				"transport: received the unexpected content-type \"application/json\"",
```
Here there was no HTTP status in the input so it should be mentioned in the error.

JNProtzman replied:
Done

JNProtzman commented on internal/transport/http_util.go:
```diff
@@ -178,15 +178,15 @@ func decodeGRPCStatusDetails(rawDetails string) (*status.Status, error) {
-	if contentTypeErr == "" {
-		errMsgs = append(errMsgs, "transport: missing content-type field")
-	} else {
+	if contentTypeErr != "" {
```
Done - I think your initialization idea is a good one.

JNProtzman commented on internal/transport/http_util.go:
```diff
@@ -178,15 +178,15 @@ func decodeGRPCStatusDetails(rawDetails string) (*status.Status, error) {
-		errMsgs = append(errMsgs, "malformed header: missing HTTP status")
-	} else {
-		errMsgs = append(errMsgs, fmt.Sprintf("%s: HTTP status code %d", http.StatusText(*(httpStatus)), *httpStatus))
+	if httpStatus != nil {
```
Done

JNProtzman commented on internal/transport/transport_test.go:
```diff
@@ -1991,6 +1991,7 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
 				Fields: []hpack.HeaderField{
 					{Name: "content-type", Value: "application/grpc"},
```
I've added a test case specifically for a missing content-type header

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1266,7 +1266,7 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		isGRPC         = !initialHeader
 		mdata          = make(map[string][]string)
-		contentTypeErr string
+		contentTypeErr = "missing content-type"
```
Actually, why not do something similar with http status?  I.e.

```go
httpStatusErr := ""
if !endStream {
	httpStatusErr = "malformed header: missing HTTP status"
}
...
	case ":status":
		if hf.Value == "200" {
			httpStatusErr = ""
			break
		}
		httpStatusErr = fmt.Sprintf(....)
...
// Delete constructHTTPErrMsg since it will just be a simple strings.Join()
```

JNProtzman replied:
Done

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -1991,17 +1991,33 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
+			// no error
+			wantStatus: status.New(
+				codes.Unknown,
+				"unexpected HTTP status code received from server: 200 (OK); missing content-type",
```
We don't want to see an error about an unexpected 200 when the only actual problem is missing content-type.

JNProtzman replied:
Done - makes total sense

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1283,6 +1283,7 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 				contentTypeErr = fmt.Sprintf("transport: received the unexpected content-type %q", hf.Value)
```
While we're here, let's finally `s/the //` please!

JNProtzman replied:
Removed `the `!

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -1991,17 +1991,33 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
+					{Name: ":status", Value: "200"},
+				},
+			},
+			// no error
```
Delete this line please

JNProtzman replied:
Done

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2066,18 +2101,24 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
 			got := ts.status
 			want := test.wantStatus
 			if got.Code() != want.Code() || got.Message() != want.Message() {
+				t.Logf("got: %s", got.Message())
+				t.Logf("got: %d", got.Code())
+				t.Logf("want: %s", want.Message())
+				t.Logf("want: %d", want.Code())
 				t.Fatalf("operateHeaders(%v); status = %v; want %v", test.metaHeaderFrame, got, want)
```
Why all the `Log`s?  Is `%v` not printing the code and message in a readable form?  Or can we delete these?

JNProtzman replied:
`%v` was not printing the code and message in a readable form. I've adjusted the error message to print a readable message and deleted the individual logs. Let me know what you think!

dfawley replied:
Oh let's add a `String()` method to `*status.Status` then, if you don't mind.  It should be trivial.  [This](https://github.com/grpc/grpc-go/blob/0257c8657362b76f24e7a8cfb61df48d4cb735d3/internal/status/status.go#L145) can then call it (`return e.e.String()`).

JNProtzman replied:
`e.e` is an `*spb.Status` which has its own `String()` implementation. In order to call the String method I'm writing, I could change the internal error type to `*Status` (see below) with very little overall change. Is this an OK change to make?

```
type Error struct {
	e *Status
}
```

dfawley replied:
Ah, right, it is a proto not a *Status in there.  Yes, this should be fine, unless something surprising comes up.  If you don't mind, please rename the field to `s` instead of `e`.

> Also, I'm not sure how to alert you (I know comments generate an email) so I've re-requested your review. I'm not sure if that's a good github practice or not, though!

Neither am I!  Whatever works for you is fine.  And feel free to ping via a comment if we don't respond after a couple days.

JNProtzman replied:
Done!

dfawley reviewed (commented):
Thanks for the changes.  The logic all looks good now, just a few minor nits.

JNProtzman commented on internal/transport/http2_client.go:
```diff
@@ -1266,23 +1267,28 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		isGRPC         = !initialHeader
 		mdata          = make(map[string][]string)
-		contentTypeErr string
+		contentTypeErr = "missing content-type"
```
Done

JNProtzman commented:
@dfawley this is ready for review when you're ready! 

I'm wondering if the test which failed is a flaky test? It passed locally (I'm running go1.16 locally) with the race flag.

dfawley commented on internal/status/status.go:
```diff
@@ -136,19 +136,23 @@ func (s *Status) Details() []interface{} {
 // GRPCStatus returns the Status represented by se.
 func (e *Error) GRPCStatus() *Status {
-	return FromProto(e.e)
+	return FromProto(e.s.s)
```
`return e.s`?  This should be safe as both types are immutable.

JNProtzman replied:
.. I totally didn't realize that was a `*Status`, great catch. Done!

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1266,23 +1267,28 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		headerError string
 	)
 
+	if !endStream {
```
Actually I believe this is incorrect, as it will not validate that a trailers-only response contains the HTTP status.  Changing this to `if initialHeader` should fix this.

dfawley replied:
Ping on this since it's still something we should fix before merging.

JNProtzman replied:
Fixed! Sorry for the delay.

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2029,10 +2029,11 @@ func (s) TestClientDecodeHeaderStatusErr(t *testing.T) {
 			},
 			wantStatus: status.New(
-				codes.Internal,
+				codes.Unknown,
```
What is the purpose of this test change?  This doesn't test the modified code AFAICT.

Can this test case list handle a new case that sets the end stream flag in the headers frame?  If so, we should be able to verify that a missing status (with all other requirements met) is detected in that scenario.  This is the case that would have failed before the change made above.

JNProtzman replied:
I reverted the test change - the message was different and I'm not sure what I was thinking. I had the tests originally set to add the end stream flag, if no `HeadersFrame` was present but I think that might have made the tests a little less explicit. Let me know what you think.

I did add a test explicitly for testing the http status missing from the headers.

dfawley commented:
Looks like `vet` is failing currently.

JNProtzman commented:
Sorry for the delay here!

dfawley reviewed (changes requested):
For the test cases, how about run each case twice: once with EndStream set and once without?  Right now it looks like it's always set, so we're missing coverage of these cases where it isn't set.

JNProtzman commented:
My laptop went belly up last night. I'm getting a new laptop in the next few days - sorry for the delay!

JNProtzman commented:
I fixed the tests to run tests for both end_stream and not. There was one test which failed - [invalid grpc status header field](https://github.com/grpc/grpc-go/blob/master/internal/transport/transport_test.go#L2000) - because we didn't validate the grpc-status in the loop any longer. Instead, we verify it if it's set after we check for end stream. I moved the check back into the loop.

There was also an end-to-end test which failed because of the same reason. The test now actually tests what it's supposed to.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1428,17 +1435,11 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
-				return
-			}
-			rawStatusCode = codes.Code(uint32(code))
+		rsc := codes.Unknown
+		if rawStatusCode != nil {
+			rsc = *rawStatusCode
```
Why not make `rawStatusCode` a non-pointer and initialize it to `codes.Unknown` instead?  Also "rsc" = "Russ Cox" :laughing: (github.com/rsc)

JNProtzman replied:
I've initialized rawStatusCode as `codes.Unknown` instead. When I wrote that I was definitely thinking about Russ haha

dfawley reviewed (approved):
LGTM, just one minor suggestion.  Thanks!

dfawley reviewed (approved):
LGTM - over to @menghanl for a second set of eyes.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1327,30 +1359,25 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		if httpStatusCode != nil {
 			var ok bool
-			code, ok = HTTPStatusConvTab[httpStatusCode]
+			code, ok = HTTPStatusConvTab[*httpStatusCode]
```
Interesting.  I just was pointed to this:

https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md

> This table is to be used only for clients that received a response that did not include grpc-status. If grpc-status was provided, it must be used


So, I think this part should only be happening if there is no grpc-status (rawStatusCode), but using rawStatusCode otherwise.  Do you want to take a shot at this?

JNProtzman replied:
I've put up a PR for if [here](https://github.com/grpc/grpc-go/pull/4606)
